src/flint/types/nmod_mpoly.pyx

```diff
@@ -0,0 +1,1451 @@
+        elif typecheck(other, nmod):
+            if other.modulus() != self.ctx.modulus():
+                raise ValueError(f"cannot subtract with different modulus {other.modulus()} vs {self.ctx.modulus()}")
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_sub_ui(res.val, (<nmod_mpoly>self).val, int(other), self.ctx.val)
+            return res
+        elif isinstance(other, int):
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_sub_ui(res.val, (<nmod_mpoly>self).val, other, self.ctx.val)
+            return res
+        else:
+            return NotImplemented
+
+    def __rsub__(self, other):
+        cdef nmod_mpoly res
+        if isinstance(other, int):
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_sub_ui(res.val, (<nmod_mpoly>self).val, other, res.ctx.val)
+            return -res
+        elif typecheck(other, fmpz):
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_sub_ui(
+                res.val,
+                (<nmod_mpoly>self).val,
+                fmpz_get_nmod((<fmpz>other).val, self.ctx.val.mod),
+                self.ctx.val
+            )
+            return -res
+        elif typecheck(other, nmod):
+            if other.modulus() != self.ctx.modulus():
+                raise ValueError(f"cannot subtract with different modulus {other.modulus()} vs {self.ctx.modulus()}")
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_sub_ui(res.val, (<nmod_mpoly>self).val, int(other), self.ctx.val)
+            return -res
+        else:
+            return NotImplemented
+
+    def __mul__(self, other):
+        cdef nmod_mpoly res
+        if typecheck(other, nmod_mpoly):
+            if (<nmod_mpoly>self).ctx is not (<nmod_mpoly>other).ctx:
+                raise IncompatibleContextError(f"{(<nmod_mpoly>self).ctx} is not {(<nmod_mpoly>other).ctx}")
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_mul(res.val, (<nmod_mpoly>self).val, (<nmod_mpoly>other).val, res.ctx.val)
+            return res
+        elif typecheck(other, fmpz):
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_scalar_mul_ui(
+                res.val,
+                (<nmod_mpoly>self).val,
+                fmpz_get_nmod((<fmpz>other).val, self.ctx.val.mod),
+                self.ctx.val
+            )
+            return res
+        elif typecheck(other, nmod):
+            if other.modulus() != self.ctx.modulus():
+                raise ValueError(f"cannot multiply with different modulus {other.modulus()} vs {self.ctx.modulus()}")
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_scalar_mul_ui(res.val, (<nmod_mpoly>self).val, int(other), self.ctx.val)
+            return res
+        elif isinstance(other, int):
+            res = create_nmod_mpoly(self.ctx)
+            nmod_mpoly_scalar_mul_ui(res.val, (<nmod_mpoly>self).val, other, res.ctx.val)
+            return res
```

Comment:
Converting the `int` to `ulong` here can overflow e.g.:
```python
In [1]: from flint import *

In [2]: ctx = nmod_mpoly_ctx(2, Ordering.lex, ['x', 'y'], 11)

In [3]: p = ctx.from_dict({(1, 2): 3})

In [4]: p
Out[4]: 3*x*y^2

In [5]: p * 100**100
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[5], line 1
----> 1 p * 100**100

File flint/types/nmod_mpoly.pyx:534, in flint.types.nmod_mpoly.nmod_mpoly.__mul__()

OverflowError: Python int too large to convert to C unsigned long
```
It also doesn't work for negative numbers:
```
In [6]: -1 * p
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[6], line 1
----> 1 -1 * p

File flint/types/nmod_mpoly.pyx:543, in flint.types.nmod_mpoly.nmod_mpoly.__rmul__()

OverflowError: can't convert negative value to ulong
```
I would say that the common case is a small integer and so it is good to use `nmod_mpoly_scalar_mul_ui` for the fast-path but it can be something like:
```python
cdef ulong o
try:
    o = other
except OverflowError:
    o = other % mod
```
There could perhaps a be a `cdef inline` function if this is something that is needed a lot. It could be a method on the context.

There is a lot of repetitive code in different methods like `__add__` for handling all of these different cases. Maybe we need a context method like `ctx.any_scalar_as_ulong(other)` that could handle the `int`, `fmpz`, `nmod`, `fmpz_mod` cases all in one.

Actually I'm not sure about `fmpz_mod`... It seems to be not supported for mixing:
```
In [1]: from flint import *

In [2]: ctx = fmpz_mod_ctx(127)

In [3]: nmod_poly([1, 2], 3) * ctx(2)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 nmod_poly([1, 2], 3) * ctx(2)

TypeError: unsupported operand type(s) for *: 'flint.types.nmod_poly.nmod_poly' and 'flint.types.fmpz_mod.fmpz_mod'

In [4]: ctx = fmpz_mod_poly_ctx(127)

In [5]: ctx([1, 2]) * nmod(3, 127)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 ctx([1, 2]) * nmod(3, 127)

TypeError: unsupported operand type(s) for *: 'flint.types.fmpz_mod_poly.fmpz_mod_poly' and 'flint.types.nmod.nmod'
```
I guess we can continue with that for `mpoly` as well.

In reality there isn't much benefit in mixing them because you should always just use `nmod` for smaller moduli and `fmpz_mod` for larger. Particularly in Python it would probably have been better if these were merged together somehow like `fq_default`.

Reply:
> Converting the `int` to `ulong` here can overflow e.g.:
> It also doesn't work for negative numbers:

This is largely intentional, though through the absence of handling rather than an explicit case. I can see why we'd want it to work as well.

> There is a lot of repetitive code in different methods like `__add__` for handling all of these different cases. Maybe we need a context method like `ctx.any_scalar_as_ulong(other)` that could handle the int, fmpz, nmod, fmpz_mod cases all in one.

I completely agree, it's turning into a huge mess and I'm not particularly happy with it but haven't opted to rip it all out in favour of something smarter yet. I certainly can, and am now considering if I could complete it today.

Just enumerating the type conversion cases for a binary operator on something like `nmod_mpoly` we have
- `int`, try casting to `ulong`
  - No overflow,  `ui` functions
  - Overflow, `int` -> `fmpz` -> `ulong` via `fmpz_get_nmod`, `ui` functions. Or `int` mod modulus in python -> `ulong`, `ui` functions
- `nmod` -> `ulong` via `.val`, ui functions
- `fmpz` -> `ulong` via `fmpz_get_nmod`, `ui` functions
- `fmpz_mod` -> `ulong` via `fmpz_get_nmod`, `ui` functions
- `nmod_mpoly`, `nmod_mpoly` functions

Some functions like `div` require going further to construct a `nmod_mpoly` and set it to a `ulong`. 

and for `fmpz_mod_mpoly`
- `int` -> `fmpz`, `fmpz` functions
- `nmod` -> `ulong` via `.val`, `ui` functions, however, flint just converts the `ulong` to a `fmpz` immediately
- `fmpz`, `fmpz` functions 
- `fmpz_mod` -> `fmpz` via `.val`, `fmpz` functions
- `fmpz_mod_mpoly`, `fmpz_mod_mpoly` functions

As it stand, the `_any_as_self` system is a step to far with respect to the optimal paths for various types. We could have a `ctx.any_as_scalar` and `ctx.scalar_as_mpoly` or similar, which depending on the type either returns the respective scalar/mpoly, provided the scalars used in the C functions are homogeneous . This can also be extended to the `fmpq_mpoly` type as well since flint just converts arguments of the `fmpz` functions to `fmpq`s immediately, so we can as well.

I'm not sure of timezones (Brisbane, UTC+10 here) but I may just go ahead and implement the above for this PR, if it works I'll port `fmpz_mpoly` and `fmpq_mpoly` over in another.

Reply:
For `nmod_mpoly` in particular is it not a bit simpler to manage scalar conversion?

We basically always want to get down to ulong (or `mp_limb_t` or something) and in the case of overflow we can use `x % mod` and then cast. This works because we know that `x % mod` will always fit the basic C type (since `mod` does).

Reply:
> For `nmod_mpoly` in particular is it not a bit simpler to manage scalar conversion?

It is indeed. I'd prefer a generic solution that works for most if not all cases across the various types and I'm writing one now. I'll have some commits up soon. 
If we have one scalar type for each of the mpoly classes then a similar approach should work for everything, given that flint seems to convert all the different types into a single one, e.g. `ulong`s into `fmpz`s for `fmpz_mod_mpoly` and `fmpq`s for `fmpq_mpoly` for example

> ```c
> // flint/src/fmpz_mod_mpoly/add_fmpz.c
> void fmpz_mod_mpoly_add_ui(
>     fmpz_mod_mpoly_t A,
>     const fmpz_mod_mpoly_t B,
>     ulong c,
>     const fmpz_mod_mpoly_ctx_t ctx)
> {
>     fmpz_t cc;
>     fmpz_init(cc);
>     fmpz_mod_set_ui(cc, c, ctx->ffinfo);
>     fmpz_mod_mpoly_add_fmpz_mod(A, B, cc, ctx);
>     fmpz_clear(cc);
> }
> 
> // flint/src/fmpq_mpoly/add_fmpq.c
> void fmpq_mpoly_add_ui(fmpq_mpoly_t A, const fmpq_mpoly_t B,
>                                            ulong c, const fmpq_mpoly_ctx_t ctx)
> {
>     fmpq_t t;
>     fmpz_init_set_ui(fmpq_numref(t), c);
>     *fmpq_denref(t) = 1;
>     fmpq_mpoly_add_fmpq(A, B, t, ctx);
>     fmpz_clear(fmpq_numref(t));
> }
> ```

I think we can as well, which will dramatically simplify things

Reply:
I've just pushed an example with `nmod_mpoly`s addition and divmod here https://github.com/flintlib/python-flint/pull/164/commits/9abc8d21a8de41788083ece5896e7deb115ffb21